client/beefy/src/worker.rs

```diff
@@ -216,77 +241,225 @@ where
-			// Authority set change or genesis set id triggers new voting rounds
-			//
-			// TODO: (grandpa-bridge-gadget#366) Enacting a new authority set will also
-			// implicitly 'conclude' the currently active BEEFY voting round by starting a
-			// new one. This should be replaced by proper round life-cycle handling.
+		trace!(target: "beefy", "🥩 new finalized block, waking up ticker");
+		// We can ignore 'buffer full' error since it means some other task already
+		// queued up a wake message.
+		let _ = self.waker_tx.try_send(());
+	}
+
+	fn handle_vote(
+		&mut self,
+		round: (Payload, NumberFor<B>),
+		vote: (Public, Signature),
+		self_vote: bool,
+	) {
+		self.gossip_validator.note_round(round.1);
+
+		let rounds = if let Some(rounds) = self.rounds.as_mut() {
+			rounds
+		} else {
+			debug!(target: "beefy", "🥩 Missing validator set - can't handle vote {:?}", vote);
+			return
+		};
+
+		let vote_added = rounds.add_vote(&round, vote, self_vote);
+
+		if vote_added && rounds.is_done(&round) {
+			if let Some(signatures) = rounds.drop(&round) {
+				self.gossip_validator.drop_round(round.1);
+
+				// id is stored for skipped session metric calculation
+				self.last_signed_id = rounds.validator_set_id();
+
+				let block_num = round.1;
+				let commitment = Commitment {
+					payload: round.0,
+					block_number: block_num,
+					validator_set_id: self.last_signed_id,
+				};
+
+				let signed_commitment = SignedCommitment { commitment, signatures };
+
+				metric_set!(self, beefy_round_concluded, block_num);
+
+				info!(target: "beefy", "🥩 Round #{} concluded, committed: {:?}.", round.1, signed_commitment);
+
+				if self
+					.backend
+					.append_justification(
+						BlockId::Number(block_num),
+						(
+							BEEFY_ENGINE_ID,
+							VersionedFinalityProof::V1(signed_commitment.clone()).encode(),
+						),
+					)
+					.is_err()
+				{
+					// just a trace, because until the round lifecycle is improved, we will
+					// conclude certain rounds multiple times.
+					trace!(target: "beefy", "🥩 Failed to append justification: {:?}", signed_commitment);
+				}
+				self.signed_commitment_sender
+					.notify(|| Ok::<_, ()>(signed_commitment))
+					.expect("forwards closure result; the closure always returns Ok; qed.");
+
+				self.set_best_beefy_block(block_num, None);
+
+				trace!(target: "beefy", "🥩 round concluded, waking up ticker");
+				// We can ignore 'buffer full' error since it means some other task already
+				// queued up a wake message.
+				let _ = self.waker_tx.try_send(());
+			}
+		}
+	}
+
+	/// Set best BEEFY block to `block_num`.
+	///
+	/// Also sends/updates the best BEEFY block hash to the RPC worker.
+	/// If the hash is not explicitly provided in `opt_block_hash`,
+	/// this function just gets it from the `Client`.
+	fn set_best_beefy_block(
+		&mut self,
+		block_num: NumberFor<B>,
+		mut opt_block_hash: Option<<<B as Block>::Header as Header>::Hash>,
+	) {
+		if opt_block_hash.is_none() {
```

Comment:
Might be worth gating the entire method to indeed check that `block_num` is an improvement over the existing one.